RFD: Support for Package URLs

alilleybrinker opened the pull request:
This adds an RFD describing a design for adding support for Package URLs to the CVE Record Format, including a deep dive on the thinking behind the design.

This RFD previously covered both Package URLs and OmniBOR Artifact IDs. After feedback from the CVE Board, that has been reduced in scope to only cover Package URLs. Another PR will be opened to cover the OmniBOR proposal.

---

- [Rendered proposal](https://github.com/CVEProject/cve-schema/blob/625cef0fe85096d070292a2ffee1119e3ab42d02/rfds/0000-expanding-software-ids.md)

alilleybrinker committed:
feat: Add RFD for Software Identifiers.

This adds an RFD describing a design for adding support for more
software identifier types to the CVE Record Format, including
a deep dive on the thinking behind the design.

Signed-off-by: Andrew Lilley Brinker <abrinker@mitre.org>

alilleybrinker commented:
First comment reserved for open questions (will be edited)

- [x] Whether to vendor the purl specification.
  - Current sense of the group: __yes, vendor the spec__
- [x] Whether to vendor the OmniBOR specification.
  - Current sense of the group: __yes, vendor the spec__
- [x] Whether to permit the "generic" and "swid" types for purl.
  - Current sense of the group: __disallow the "generic" and "swid" types__
- [x] Whether to limit the number of software identifiers in a single CVE record.
  - Current sense of the group: __do not limit the number, rely on overall CVE record size limit__
- [x] Whether to permit versions directly in a purl. They are currently allowed in a CPE, but by convention if provided in a CPE then version constraint fields should not be used.
  - Current sense of the group: __match existing CPE behavior__
- [x] Whether to use the `vers` spec for specifying version constraints instead of reusing the `cpeMatch` version fields.
   - Current sense of the group: __do not use the `vers` spec__
- [x] Whether to add more software identifiers to the `affected` array rather than modifying the `cpeApplicability` object.
  - Current sense of the group: __use the `affected` array__

__Issues considered resolve are marked with a checkmark. If you believe an issue is not resolved, please raise it in a comment below.__

alilleybrinker commented:
I've now opened PRs reflecting an alternate design to the one proposed in this RFD, #409 and #410. If the QWG's consensus is to advance with those designs, I will close this RFD PR and open a new PR with an alternate RFD describing those designs in detail.

darakian commented:
Having both of these approaches open may not have been the best idea (for me anyway :smile: ). I've added my comment about how to encode version ranges in the other PR though I suspect that would also be relevant here as well.

I will also be totally honest that I have not dug into Omnibor yet and I've been stuck in a loop where a QWG meeting happens and Omnibor is mentioned, I think `Oh shoot I forgot to read up on that`, I take a quick look at it and get intimated by the spec, put it down and swear that I'll get back to it ~tomorrow~,~the next day~, ~the day after that~,~etc....~, and then there's another QWG. I won't be read up on OmniBOR by tomorrow's QWG, but please yell at me if I'm not y next weeks meeting :+1: 

So, with that personal failing out of the way I'll continue to be self centered and pose the question; would it make sense to drop both purl and omnibor from this RFD and discuss the idea of how we should consider software identifiers more generally? I've been waffling on the idea of proposing a set of what I'll call `the identifiers that github uses` which are similar to a subset of purls or not. I lean more toward not at this point, but I do think we should consider the generic question of `how do we assess a new identifier` should `Omnibor2` or `cpeButBetterThisTime`, or `whatever` comes along. If nothing else this vetting process could be used as a way to provide feedback upstream. Success metric could also be easier to consider if the RFD is broken up as well. That said, I'm happy to hear pushback on that too as there really are not many software identifiers worth considering today and it may just be easier to have the full conversation all at once.

---

With respect to this RFD as is; I am onboard with the general direction of it. I think I understand the synonym problem to be one that is more operational than schema design and I would suggest that allowing for more methods to capture affected product information may enable/embolden more CNAs to publish affected product information period. In my opinion an indicator of success could be a rise in the proportion of CVEs published with valid affected product information populated. This could also be broken out per id (and for the structure itself) to indicate success for each.

On the `Related Issues or Proposals` section; this could be a case of `yes and`. CPE has its issues today and is currently bottlenecked with NIST as a centralized naming authority. I see no reason why CPE supporters couldn't continue development of CPE and federate the namespace based on a per-vendor basis or whatever and that be compatible with the adoption of purls/omnibors. I suspect that orgs will pick whichever namespace fits them and their needs best and IMO the spec should be equipped to accept whatever high quality identifiers can be produced.

alilleybrinker commented:
@darakian, for reading up on OmniBOR, the project website has a more accessible introduction to how the identifiers work: https://omnibor.io/docs/artifact-ids/

As for the question of whether to split this RFD into parts: one advancing a general set of provisions for how new software IDs should be incorporated, and then others advancing specific software IDs to incorporate. My team is open to doing that, but didn't as the initial ask because we were concerned it would be too granular for the QWG. In particular, we felt that offering concrete examples with real-world identifiers helps crystallize understanding of the trade-offs of any particular design in a way that a purely abstract proposal can't.

For the success metric question, I agree there's value in assessing adopting via a statistical analysis of uptake of the new fields or general enrichment of identity information in CVE records, though we didn't choose that as the go-to metric because we wanted to leave room to also consider adoption by CVE consumers, which is fuzzier to measure and thus easier to ignore when assessing success. That said, I'd love to see the kind of analysis you mention done after maybe six months post-adoption.

Finally, on the related issues, these are more listed to identify problems we are explicitly not solving in the RFD but which the QWG could take up and pursue outside of the RFD. I agree that there's interest and a need to advance some form of improvement to CPE, likely via federation, to let it scale beyond the limits of NIST's resources; we just don't solve that question here.

alilleybrinker commented:
@darakian we could amend the RFD text to include a clear subsection which describes that the design proposed here for adding Package URLs and OmniBOR Artifact Identifiers is intended as a template for addition of any future identifier types, which may help satisfy the desire for a clear reference-point on how to add those types in future proposals, without needing to split the RFD into multiple separate documents.

darakian commented:
> @darakian, for reading up on OmniBOR, the project website has a more accessible introduction to how the identifiers work: https://omnibor.io/docs/artifact-ids/

Thank you, thank you!

> As for the question of whether to split this RFD into parts: one advancing a general set of provisions for how new software IDs should be incorporated, and then others advancing specific software IDs to incorporate. My team is open to doing that, but didn't as the initial ask because we were concerned it would be too granular for the QWG.

Ya, that's totally fair. I was very much waffling back and forth on if to raise the issue or not.

> @darakian we could amend the RFD text to include a clear subsection which describes that the design proposed here for adding Package URLs and OmniBOR Artifact Identifiers is intended as a template for addition of any future identifier types, which may help satisfy the desire for a clear reference-point on how to add those types in future proposals, without needing to split the RFD into multiple separate documents.

That could work for sure. 👍

alilleybrinker committed:
feat: Base the Software ID RFD on the affected array

This rewrites the core content of the RFD to base the
proposed new fields on the `affected` array instead of basing
them on the `cpeApplicability` object as the prior version of
the RFD did. The motivation and outcomes are generally unchanged,
but the specifics of the proposed edits are now different.

Signed-off-by: Andrew Lilley Brinker <abrinker@mitre.org>

alilleybrinker commented:
@darakian, thanks! I've amended the RFD to be based on the `affected` array and to include more explicit commentary on how it is intended to function as a template for the inclusion of future identifier types.

Chris-Turner-NIST commented:
Finally getting a moment to read through this and realized that this may not have been covered in discussions yet...  

If the intent is to create more generic places for various identifiers, it would make sense that part of this proposal should include deprecating the existing `cpes` array and include a new property (`cpeMatchString`?) that aligns with the approach proposed for PURL and OmniBOR.  

I recognize that this would create two locations for CPE related data due to the current support for `hasCPEApplicability`, however, it would be a step in the right direction of normalizing the current structures and methodologies available within the affected array.

alilleybrinker commented:
@Chris-Turner-NIST I agree that it would be good both to eventually deprecate the `cpes` array and to introduce a field for CPEs similar to the support added for OmniBOR and Package URLs in this RFD. However, we purposefully omitted that issue in this RFD for a couple of reasons:

1. Deprecations are harder to justify, would likely take longer to reach consensus on how to handle a deprecation proposal.
2. The `cpeApplicability` block, added last year as an NVD-compatible mechanism for adding CPEs which is also semantically clearer than the `cpes` field in the `affected` array's `product` objects, complicates the story around "where CPEs go" in an CVE record. If the `cpes` field were deprecated and a new `cpeMatch` (or some other name) field were added in the same object, CNAs would still be presented with 3 places to put CPEs, with one deprecated.

All this to say, I fully endorse improving and simplifying handling of CPEs in the record format, and my personal preference is to do exactly what you propose. I just think it would make the most sense in a follow-up RFD.

alilleybrinker commented:
@Chris-Turner-NIST, I've opened an Issue recommending the creation of an RFD for improving CPE handling, based on your comment here. Happy for any additional input you may have on that! https://github.com/CVEProject/cve-schema/issues/421

andrewpollock commented on rfds/0000-expanding-software-ids.md:
```diff
@@ -0,0 +1,484 @@
+identifier. The reliance on this central dictionary means that the issuance of
+new CPEs for vendors or products not present in the dictionary requires NIST
+to update the dictionary to support them. Per NIST's procedures, to make the
+workload of maintaining the dictionary manageable, new entries in the dictionary
```
Nit: as I understand it, this is a widely held myth. One can preemptively request CPEs be added to the CPE Dictionary. @Chris-Turner-NIST could you please confirm?

alilleybrinker replied:
Confirmed in today's QWG meeting that this is a myth. The RFD has been amended accordingly.

andrewpollock commented on rfds/0000-expanding-software-ids.md:
````diff
@@ -0,0 +1,484 @@
+```json
+"affected": [
+    {
+        "artifactID": "gitoid:blob:sha256:9f64df92367881be21e23567a31a8ce01994d98b69d28917b5c132ce32a8e6c8",
````
Could you talk to how a downstream consumer would divine the Git repository this hash relates to?

alilleybrinker replied:
It doesn't relate to a Git repository. The Git Object ID construction is used for easy cross-reference into Git repositories, but the identifier is based only on the artifact itself. If a consumer has the artifact, they can check its Artifact ID against the Artifact ID in a CVE record to determine if they're vulnerable.

alilleybrinker replied:
Here's the OmniBOR documentation for how Artifact IDs are constructed, for reference: https://omnibor.io/docs/artifact-ids/

It's just a prefix string encoding the length of the artifact in bytes, and then the contents of the artifact, with newlines normalized to Unix-style, fed into a SHA-256 hasher.

andrewpollock commented on rfds/0000-expanding-software-ids.md:
```diff
@@ -0,0 +1,484 @@
+Package URL, but again lack a well-defined construction to ensure consistency
+and easy automated cross-referencing between datasets. The `collectionURL`
+field provides many example, but is not constrained beyond requiring a valid
+URI. The `packageNamem` field is entirely unconstrained. Similarly to the first
```
Typo fix

```suggestion
URI. The `packageName` field is entirely unconstrained. Similarly to the first
```

andrewpollock commented on rfds/0000-expanding-software-ids.md:
```diff
@@ -0,0 +1,484 @@
+matching of CVE records, but not for automated processing of applicability at
+scale.
+
+## Proposed Solution
```
In terms of "alternatives considered"...

Why not address the lack of convention to the `vendor`/`product` and/or `collectionURL`/`packageName` fields rather than proposing adding a whole new field?

e.g. `vendor` and `product` could by convention be required to map the vendor and product of a CPE, and then be  able to be validated against the CPE Dictionary. The expectations for `collectionURL` and `packageName` could similarly be defined.

alilleybrinker replied:
Because those are restrictive changes that would either constrain the format in a way that's not backwards-compatible or otherwise be merely normative and thereby of questionable value. Also in the case of OmniBOR there's no clear precedent value to use either.

alilleybrinker committed:
fix: Correct `packageName` typo.

Co-authored-by: Andrew Pollock <andrewpollock@users.noreply.github.com>

alilleybrinker committed:
feat: Make RFD *only* about Package URLs

After feedback from the CVE Board, this RFD is being split into two,
with this part covering *only* the addition of support for Package
URLs, and the other part (to be opened as a separate RFD) covering
support for OmniBOR Artifact IDs.

This commit does the split, and makes no edits besides removing the
OmniBOR pieces and smoothing out the remaining language where
necessary.

Signed-off-by: Andrew Lilley Brinker <alilleybrinker@gmail.com>

feat: ensure new IDs can't fulfill "identifier-like" requirement

Per discussion in the QWG, this amends the RFD to clarify that the new
identifier fields being proposed are not able to fulfill the "identifier-like"
requirement in the `product` object inside the `affected` array. While this
may be changed in the future, for today it is the easiest path forward for
CVE data consumers, who could adopt the new fields if _desirable_ but would
not be obligated to do so.

Signed-off-by: Andrew Lilley Brinker <alilleybrinker@gmail.com>

alilleybrinker committed:
feat: ensure new IDs can't fulfill "identifier-like" requirement

Per discussion in the QWG, this amends the RFD to clarify that the new
identifier fields being proposed are not able to fulfill the "identifier-like"
requirement in the `product` object inside the `affected` array. While this
may be changed in the future, for today it is the easiest path forward for
CVE data consumers, who could adopt the new fields if _desirable_ but would
not be obligated to do so.

Signed-off-by: Andrew Lilley Brinker <alilleybrinker@gmail.com>

alilleybrinker commented:
> [!note]
> ## Final Comment Period
> A __Final Comment Period__ (FCP) has been called for this proposal. This is a final opportunity to raise new concerns with the proposal.
> 
> The FCP will close at 2pm PDT / 5pm EDT July 3rd, at the end of the Quality Working Group Meeting.

alilleybrinker commented:
> [!note]
> ## Final Comment Period Has Closed
> The __Final Comment Period__ (FCP) for this proposal has closed, and the proposal has been __accepted__ by the QWG.
> 
> Per the RFD process rules, it will now advance to the CVE Board for consideration. The Board will make the final determination as to whether to adopt or reject the proposal.

alilleybrinker commented:
After deliberation with the CVE Board, we've decided to split this RFD into two parts: one for Package URLs and one for OmniBOR Artifact IDs.

The plan will be to likely proceed forward with the Package URL RFD, while continuing to work on improvements to the OmniBOR Artifact ID RFD.

For the OmniBOR Artifact ID RFD, the design will be revisited, with an eye toward handling fine-grained identifiers like OmniBOR Artifact IDs in a distinct manner from the way course-grained identifiers are used in CVE Records today.

alilleybrinker commented:
Gah, seems I errored in my commit-making and rebased the last commit before the new one I just added. To be clear: that commit has not been edited, this only adds a new commit which splits out the OmniBOR parts of the original RFD. This RFD is now solely about Package URLs; a separate PR will be opened for OmniBOR.

alilleybrinker committed:
chore: rename RFD file to match new title

Since this RFD is now Package URL specific, this renames the RFD
file to reflect the new title.

Signed-off-by: Andrew Lilley Brinker <abrinker@mitre.org>

alilleybrinker commented:
From today's QWG discussion: there was consensus on the value of pursuing a validation library for CVE records, given that with the introduction of Package URLs it will no longer be sufficient to check that CVE Records pass JSON schema validation. This would need to be a discussion with the AWG, which @ccoffin has taken an action to start.

pombredanne commented:
@alilleybrinker re:
> From today's QWG discussion: there was consensus on the value of pursuing a validation library for CVE records, given that with the introduction of Package URLs it will no longer be sufficient to check that CVE Records pass JSON schema validation. This would need to be a discussion with the AWG, which @ccoffin has taken an action to start.

See the work we are doing to bring open source validation at https://github.com/package-url/purl-spec/discussions/296 repasted here for reference:

> 
> I am working on some extensive tooling and data for PURL validation!
> 
> See these many issues to track the work specifically on PURL validation:
> - https://github.com/orgs/aboutcode-org/projects/52/views/22?filterQuery=projectid%3Apurl-next+issueid%3Apurl-next-m3*%2Cpurl-next-m5*
> - https://github.com/orgs/aboutcode-org/projects/52/views/10
> - https://github.com/orgs/aboutcode-org/projects/52/views/21
> - https://github.com/orgs/aboutcode-org/projects/54
> 
> In particular:
> - validate PURL overall https://github.com/aboutcode-org/purl-spec/issues/43
> - create a data-driven offline validator in Rust/Go/Python: 
>   - https://github.com/aboutcode-org/purl-validator/issues/18
>   - https://github.com/aboutcode-org/purl-validator/issues/20
> - same but online: https://github.com/aboutcode-org/purldb/issues/643
> - adding extended validation in Python: https://github.com/aboutcode-org/purl-spec/issues/44
> - validate SBOMs https://github.com/aboutcode-org/scancode.io/issues/1597 
> - help validate accuracy of scanner/tools, including for binaries:
>   - https://github.com/aboutcode-org/purl-spec/issues/48
>   - https://github.com/aboutcode-org/purl-spec/issues/66
>   - https://github.com/aboutcode-org/scancode.io/issues/1770
>   - https://github.com/aboutcode-org/purl-spec/issues/61
> 
> Also, help major FOSS foundations and registries adopt/use the tools:
> - https://github.com/aboutcode-org/purl-spec/issues/56
> 
> (separately crates.io now displays PURL on each crate's page)
> 
> @andrewpollock so as you can se there is a lot of things planned. A shit load of work to do! Help is mucho wanted :angel: 
> 

_Originally posted by @pombredanne in https://github.com/package-url/purl-spec/discussions/296#discussioncomment-14114066_

pombredanne commented:
@alilleybrinker tell me if I can join a future QWG call to present the gist of the work

alilleybrinker commented:
> @alilleybrinker tell me if I can join a future QWG call to present the gist of the work

I'd certainly be interested to hear about it. CCing @ccoffin, one of the QWG Co-Chairs, who maintains the agendas.

alilleybrinker commented:
My team has drafted some initial guidance for CNAs and CVE Consumers, respectively, trying to answer key questions for how to use the new field: https://gist.github.com/alilleybrinker/5c38a0e176482f475b809b17156d5a5f

As the QWG/Board get into prepping for this to be released in (presumably) 5.2.0 of the Record Format, this could be a useful starting point for materials to share with stakeholders about the change.

ccoffin commented:
> @alilleybrinker tell me if I can join a future QWG call to present the gist of the work

Hi Phillipe!

It would be great to hear from you regarding the new work! Our normal meeting time is 3-4 PM EST, but i think last time we setup a 10-11 AM EST slot that was more compatible with your time zone. Let me know what you prefer.

Chris

pombredanne commented:
@ccoffin either time works for me. Next week is busy as I will be at https://osseu2025.sched.com but your regular time can work even then.

ccoffin commented:
@pombredanne OK sounds good! Shall we go ahead and plan for you to present in the regularly scheduled QWG meeting on Aug 28?

alilleybrinker committed:
chore: Use RFD #2

Signed-off-by: Andrew Lilley Brinker <abrinker@mitre.org>

pombredanne commented:
@ccoffin Looking forward to the discussion today!
